internal/secret-storage-backend/server.go

```diff
@@ -0,0 +1,315 @@
+package secretstoragebackend
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+
+	"capact.io/capact/internal/ptr"
+	pb "capact.io/capact/pkg/hub/api/grpc/storage_backend"
+	"github.com/pkg/errors"
+	tellercore "github.com/spectralops/teller/pkg/core"
+	"go.uber.org/zap"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// AdditionalParameters holds Secret storage backend specific parameters.
+type AdditionalParameters struct {
+	Provider string `json:"provider"`
+}
+
+var _ pb.StorageBackendServer = &Handler{}
+
+// Handler handles incoming requests to the Secret storage backend gRPC server.
+type Handler struct {
+	pb.UnimplementedStorageBackendServer
+
+	log *zap.Logger
+
+	providers map[string]tellercore.Provider
+}
+
+const (
+	lockedByField        = "locked_by"
+	firstResourceVersion = 1
+)
+
+var (
+	// NilRequestInputError describes an error with an invalid request.
+	NilRequestInputError = status.Error(codes.InvalidArgument, "request data cannot be nil")
+)
+
+// NewHandler returns new Handler.
+func NewHandler(log *zap.Logger, providers map[string]tellercore.Provider) *Handler {
+	return &Handler{
+		log:       log,
+		providers: providers,
+	}
+}
+
+// GetValue returns a value for a given TypeInstance. It returns nil as value if a given secret is not found.
+func (h *Handler) GetValue(_ context.Context, request *pb.GetValueRequest) (*pb.GetValueResponse, error) {
+	if request == nil {
+		return nil, NilRequestInputError
+	}
+
+	provider, err := h.getProviderFromAdditionalParams(request.AdditionalParameters)
+	if err != nil {
+		return nil, err
+	}
+
+	key := h.storageKeyForTypeInstanceValue(provider, request.TypeinstanceId, request.ResourceVersion)
+	entry, err := h.getEntry(provider, key)
+	if err != nil {
+		return nil, err
+	}
+
+	var value []byte
+	if entry.IsFound {
+		value = []byte(entry.Value)
+	}
+
+	return &pb.GetValueResponse{
+		Value: value,
+	}, nil
```

Comment:
IMO on the Local Hub side it will be good to get info if data was not found and propagate it to user.

Now you return `nil` for `value` but IMO it's not readable from API perspective.

Is it an approach suggested on gRPC calls?

--- 
after some reading IMO we should return normal code instead of playing with `nil` values. In a result, on client side you have nice error:
```
  code: 5,
  details: 'TypeInstance "123" in revision 1 was not found',
```
and we can easily do `error.code == grpc.status.NOT_FOUND` _(js code)_

**suggestion**
```suggestion
	if !entry.IsFound {
		return nil, status.Error(codes.NotFound, fmt.Sprintf("TypeInstance %q in revision %d was not found", request.TypeinstanceId, request.ResourceVersion))
	}

	return &pb.GetValueResponse{
		Value: []byte(entry.Value),
	}, nil
```
based on the above comment, additionally, it's more readable.

Reply:
Yes, we can definitely return errors 👍